library/src/main/java/org/codice/security/sign/Decoder.java

```diff
@@ -48,38 +54,51 @@ public static String decodePostMessage(String message) throws IOException {
    * @param message - SAML Redirect message
    * @return - decoded message
    */
-  public static String decodeAndInflateRedirectMessage(String message) throws InflationException {
+  public static String decodeAndInflateRedirectMessage(String message) throws DecoderException {
     String urlDecoded;
+
     try {
       urlDecoded = URLDecoder.decode(message, StandardCharsets.UTF_8.name());
     } catch (UnsupportedEncodingException e) {
-      throw new InflationException(InflationException.InflErrorCode.ERROR_DECODING);
+      throw new DecoderException(DecoderException.InflErrorCode.ERROR_URL_DECODING);
     }
+
     if (urlDecoded.matches("[ \\t\\n\\x0B\\f\\r]+")) {
-      throw new InflationException(InflationException.InflErrorCode.LINEFEED_OR_WHITESPACE);
+      throw new DecoderException(DecoderException.InflErrorCode.LINEFEED_OR_WHITESPACE);
     }
 
-    byte[] deflatedValue = Base64.getDecoder().decode(urlDecoded.getBytes(StandardCharsets.UTF_8));
+    // this method is only here to try and catch a base64 decoding error
+    try {
+      new String(Base64.getDecoder().decode(urlDecoded.getBytes(StandardCharsets.UTF_8)),
+          StandardCharsets.UTF_8.name());
+    } catch (UnsupportedEncodingException e) {
+      throw new DecoderException(InflErrorCode.ERROR_BASE64_DECODING);
+    }
+
+    byte[] base64Decoded = Base64.getDecoder().decode(urlDecoded.getBytes(StandardCharsets.UTF_8));
```

Comment:
❓ Why are you doing this here when you're doing it in the try block above?

Reply:
The byte array is needed for further computation below.
The try block isn't *really* needed, however it is used to catch a base64 decoding exception and throw the correct SAML Compliance error.

Reply:
I'm fine with all of that, but you're decoding the array twice which is inefficient.

Reply:
Fixed it!